src/engine/automaton/automaton.ts

```
@@ -93,6 +93,105 @@ export class Automaton implements IAutomaton {
   }
 
+  isDeterministic(): boolean {
+    const alphabet = new Set<string> ([]);
+    const stackAlphabet = new Set<string> ([]);
+
+    if (this.automatonType == AutomatonType.PDA) {
+      stackAlphabet.add(INITIAL_STACK_SYMBOL);
+    }
+
+    // goes through the entire deltaFunctionMatrix and add all the alphabet symbol to the alphabet (same for stack symbols)
+    for (const leftSymbol in this.deltaFunctionMatrix) {
+      for (const rightSymbol in this.deltaFunctionMatrix[leftSymbol]) {
+        const edges = this.deltaFunctionMatrix[leftSymbol][rightSymbol];
+        for (const edge of edges) {
+          alphabet.add(edge.inputChar);
+          if (this.automatonType == AutomatonType.PDA && edge instanceof PDAEdge) {
+            stackAlphabet.add(edge.readStackChar);
+            for (const symbol of edge.writeStackWord) {
+              stackAlphabet.add(symbol);
+            }
+          }
+        }
+      }
+    }
+
+    switch (this.automatonType) {
+      case AutomatonType.FINITE:
+        // if there is Epsilon it is non-deterministic
+        if (alphabet.has(EPSILON)) {
+          return false;
+        }
+        // if any state has no edges in deltafunction, automaton is non-deterministic
+        if (this.states.length != Object.keys(this.deltaFunctionMatrix).length) {
+          return false;
+        }
+        // now, for each node we go through the deltaMatrix a second time and for each edge we check
+        // if all the symbols are present and if the number of edges
+        // equals this number, therefore we check if each edge is used only once
+        for (const leftSymbol in this.deltaFunctionMatrix) {
+          const found = new Set<string> ([]);
+          let edgeNum = 0;
+
+          for (const rightSymbol in this.deltaFunctionMatrix[leftSymbol]) {
+            const edges = this.deltaFunctionMatrix[leftSymbol][rightSymbol];
+            for (const edge of edges) {
+              found.add(edge.inputChar);
```

The PDA case checks if the edges have type PDAEdge. I think a similar check would be good here.

---

yea I agree. I also added throw error in case of wrong edge instead of just returning false.